### PR TITLE
Add proguard to the android gate

### PR DESCRIPTION
**Pull request by timtay-microsoft**
[Proguard](https://www.guardsquare.com/en/products/proguard) is an android specific minifying/obfuscation tool that some users apply to their code. These changes will make it so that our android e2e test runs will be done on minified versions of our SDK so that users don't run into issues like #1032

### CODE REVIEW

File: device/iot-device-samples/android-sample/app/proguard-rules.pro [hunk 1]
```diff
@@ -15,3 +15,26 @@
...
+#-keep class com.microsoft.azure.sdk.iot.service.ImportMode { *; }
+#-keep class com.microsoft.azure.sdk.iot.service.AuthenticationMechanism { *; }
+
+# These are the recommended exclusions when using the Device Provisioning Service device client SDK.
+# They are commented out because this sample only uses the IoT Hub device client SDK
+#-keep class com.microsoft.azure.sdk.iot.provisioning.device.internal.parser.* { *; }
+
+# These are the recommended exclusions when using the Device Provisioning Service service client SDK.
+# They are commented out because this sample only uses the IoT Hub device client SDK
+#-keep class com.microsoft.azure.sdk.iot.provisioning.service.configs.* { *; }
```
Comment (timtay-microsoft) [inline device/iot-device-samples/android-sample/app/proguard-rules.pro:40]: These "-keep class x.y.z" lines just mean that proguard should deliberately **not** minify the class or package in question

File: device/iot-device-samples/android-sample/build.gradle [hunk 2]
```diff
@@ -6,11 +6,15 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.1.4'
+        classpath 'com.android.tools.build:gradle:4.1.2'
```
Comment (timtay-microsoft) [inline device/iot-device-samples/android-sample/build.gradle:9]: Out gradle versions were a bit out of date, so now seems like an appropriate time to bump them

File: device/iot-device-samples/android-sample/app/proguard-rules.pro [hunk 3]
```diff
@@ -15,3 +15,26 @@
 #-keepclassmembers class fqcn.of.javascript.interface.for.webview {
 #   public *;
 #}
+
+# JSON parser classes don't mix well with minification because JSON parser classes use reflection
+# based on the name of the field to create the JSON, and minification changes those names. As a result
+# all the message payload serializing and deserializing breaks. These lines explicitly exclude
+# the packages where we keep our JSON parsing classes from minification to avoid this issue
```
Comment (abhipsaMisra) [inline device/iot-device-samples/android-sample/app/proguard-rules.pro:22]: Interesting find!

File: device/iot-device-samples/android-sample/app/proguard-rules.pro [hunk 4]
```diff
@@ -15,3 +15,26 @@
...
+-keep class com.microsoft.azure.sdk.iot.deps.serializer.* { *; }
+-keep class com.microsoft.azure.sdk.iot.deps.twin.* { *; }
+-keep class com.microsoft.azure.sdk.iot.device.edge.MethodRequest { *; }
+-keep class com.microsoft.azure.sdk.iot.device.hsm.parser.ErrorResponse { *; }
+-keep class com.microsoft.azure.sdk.iot.device.hsm.parser.SignRequest { *; }
+
+# These are the recommended exclusions when using the IoT Hub service client SDK. They are commented
+# out because this sample only uses the IoT Hub device client SDK
+#-keep class com.microsoft.azure.sdk.iot.service.ImportMode { *; }
```
Comment (abhipsaMisra) [inline device/iot-device-samples/android-sample/app/proguard-rules.pro:31]: This is very useful info, but is the device client sample the right place for this? Do we have service client/ provisioning etc android samples where we could place these lines instead?

**Review by abhipsaMisra — APPROVED** [review]
cool!

File: device/iot-device-samples/android-sample/app/proguard-rules.pro (hunk 4 above)
Comment (timtay-microsoft) [reply device/iot-device-samples/android-sample/app/proguard-rules.pro:31]: We don't have an android sample for service client or for our provisioning clients, so this is the best place for now

**Pull request merged**